### DISCUSSION TITLE
RunBeforeExport/RunAfterExport: calling a procedure from an Access add-in

### DISCUSSION THREAD

**Discussion by josef-poetzl** [joyfullservice/msaccess-vcs-addin#493] [Ideas]
Allow to start a procedure from an Access add-in with export option _Run Sub Before Export_.
Example to start automated tests: %appdata%\Microsoft\AddIns\AccUnitLoader.AutomatedTestRun

Possible implementation:
https://github.com/josef-poetzl/msaccess-vcs-addin/commit/7921149254721f9f6c4f88615886d5a342a4bb0f

**Comment by josef-poetzl** [comment]
Note: 

Since Application.Run does not forward the error description, I now use this interface:
* `Public Function ProcedureNameInAddIn() as String`
   => if len(return values) > 0 => show text
       optional with eErrorLevel code like "Error: Error description" or "Warning: warning message to show"
* `Public Function ProcedureNameInAddIn() as Boolean`
       => if false then Error (eelCritical)

Details:
https://github.com/josef-poetzl/msaccess-vcs-addin/commit/fc902e42b360ad6141d98e98d3aa072efc32677c

See modification in action:
https://www.youtube.com/watch?v=pGJr5YjuVn4

**Comment by josef-poetzl** [comment]
New Version (4.1.0): https://github.com/josef-poetzl/msaccess-vcs-addin/releases/tag/v4.1.0-call-add-in-procedure

Are you interested in implementing the functionality of calling an add-in procedure in the original?
Note: The advantage is that you don't have to have cleanup functionality or similar in the main application.

**Reply by joyfullservice** [reply]
Sorry, I guess I must have missed your original discussion from last year. Yes, I do like the idea of being able to surface a return value in the console/log display.

As food for thought, we could even take this a step further and expose an API in the add-in that would allow the external process to post incremental progress back to the log window and even increment the progress bar. When the external process is complete, it would call the add-in to resume the process, and the add-in would pick back up from there.

Essentially, you would have three ways to utilize a hook.
1. Simply call an external function, like what you have right now.
2. Call a function, but also handle a return value, posting a message in the log window. (Like what you are proposing.)
3. Call a function, and release processing to the function (or add-in) to do it's work, and handle log posts sent back to the add-in. The external process would then be responsible to call a `.Resume` function in the add-in, signalling that the external processing is complete, and optionally passing a status code or error level.

For something more advanced like unit testing, I could see that third option as providing lots of flexibility for a more integrated build/test experience. If a developer is using a very complex `AfterBuild` process, it might be nice to see periodic messages posted in the log window as various steps are completed.

**Reply by josef-poetzl** [reply]
An interface for adding log lines is already available: ```VCS.AddLog```.

**Comment by josef-poetzl** [comment]
Next step: Run several processes at RunBeforeExport/RunAfterExport
Video (youtube): https://youtu.be/WpU58qZiZsY

Without an add-in call, this feature requires little code customization (see: https://github.com/joyfullservice/msaccess-vcs-addin/compare/dev...josef-poetzl:msaccess-vcs-addin:dev-call-multiple-procedures).
But it is only useful with the add-in procedure call.

Note: I separate the procedures with `:` (VBA syntax)

in TextBox:
![export-multi-proc-textbox](https://github.com/user-attachments/assets/ba391666-a20f-47f9-bfb7-c527584dd544)

as list:
![export-multi-proc-list](https://github.com/user-attachments/assets/e66dd1aa-81ce-41ad-8d3c-78ea61e66f20)

**Comment by josef-poetzl** [comment]
Update: implemented in v5

Step-by-step modifications to follow:

1. ErrorInfo class
   https://github.com/joyfullservice/msaccess-vcs-addin/compare/dev...josef-poetzl:msaccess-vcs-addin:dev-v5-01-ErrorInfo
2. Procedure runner class
   https://github.com/josef-poetzl/msaccess-vcs-addin/compare/dev-v5-01-ErrorInfo...josef-poetzl:msaccess-vcs-addin:dev-v5-02-runprocedure
3. Allow parameters in external procedures
   https://github.com/josef-poetzl/msaccess-vcs-addin/compare/dev-v5-02-runprocedure...josef-poetzl:msaccess-vcs-addin:dev-v5-03-runprocedure-vcs
4. Config 1..n procedures to call befort/after export
   https://github.com/josef-poetzl/msaccess-vcs-addin/compare/dev-v5-03-runprocedure-vcs...josef-poetzl:msaccess-vcs-addin:dev-v5-04-runprocedure-multi
5. call procedures from add-in
   https://github.com/josef-poetzl/msaccess-vcs-addin/compare/dev-v5-04-runprocedure-multi...josef-poetzl:msaccess-vcs-addin:dev-v5-05-runprocedure-add-in
6. get log data from external procedure --- !! new items in eErrorLevel !!
   https://github.com/josef-poetzl/msaccess-vcs-addin/compare/dev-v5-05-runprocedure-add-in...josef-poetzl:msaccess-vcs-addin:dev-v5-06-log-external-proc
7. UI: insert multiple procs in list mode 
  https://github.com/josef-poetzl/msaccess-vcs-addin/compare/dev-v5-06-log-external-proc...josef-poetzl:msaccess-vcs-addin:dev-v5-07-runprocedures
